### PR TITLE
Geal/clean up dedup on cancel

### LINKED ISSUE #754: improved async cancellation handling will require more hashmap management
The fix (#752 ) for #550 resolves the immediate problem of the router hanging when cancelled tasks are re-encountered. However, the fix is "lazy", in that the HashMap which implements our wait_map is only cleaned up when the same request if re-submitted.

IRL a lot of different requests may be submitted and not cleaned up (if the requests are cancelled) which could lead to memory bloat.

We'd like to avoid that problem by managing the HashMaps and removing inaccessible broadcasters when we detect they are no longer required.

There are a variety of approaches we can use to solve this problem, including:
 - time based LRU cache with filter matching
 - periodic sweeps of the HashMap to remove stale entries
 - stale entry detection/removal at cancellation time

### PR DESCRIPTION
**Pull request by Geal**
Fix #754

This takes care of cleaning up the wait map by using a oneshot channel as drop signal. This is a pattern that I used a lot in [pulsar-rs](https://github.com/wyyerd/pulsar-rs) to perform async tasks on drop

### CODE REVIEW

**Comment by Geal** [comment]
hmm, looks like I created this branch from the wrong one :sweat_smile: 

I'll clean that up. The relevant code is in https://github.com/apollographql/router/pull/758/commits/f261ff944f53a517ad07cd48f8d1156e5c63a63f

File: apollo-router-core/src/layers/deduplication.rs [hunk 1]
```diff
@@ -78,29 +67,25 @@ where
...
                     let http_request = request.http_request.clone();
-                    let res = match service.ready_oneshot().await {
-                        Ok(mut s) => s.call(request).await,
-                        Err(e) => {
-                            // Clean up wait map if ready_oneshot failed
+                    let res = {
+                        // when _drop_signal is dropped, either by getting out of the block, returning
+                        // the error from ready_oneshot or by cancellation, the drop_sentinel future will
+                        // return with Err(), then we remove the entry from the wait map
+                        let (_drop_signal, drop_sentinel) = oneshot::channel::<()>();
```
Comment (BrynCooke) [inline apollo-router-core/src/layers/deduplication.rs:80]: Nit: could be a mutex rather than a channel.

**Review by BrynCooke — APPROVED** [review]
LGTM

**Review by garypen — CHANGES_REQUESTED** [review]
We should also update: https://github.com/apollographql/router/blob/main/apollo-router-core/src/cache.rs
to maintain consistency.

**Comment by Geal** [comment]
`cache.rs` is a bit tricky with the two locks, I'll send it in a separate PR so we can take the time to study it

**Comment by Geal** [comment]
can we agree to merge this one at least?

**Review by garypen — APPROVED** [review]

**Pull request merged**